drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,863 @@
+
+	/* Config clocks and analog filter */
+	DL_I2C_setClockConfig(config->base, &config->gI2CClockConfig);
+	DL_I2C_disableAnalogGlitchFilter(config->base);
+
+	/* Configure Controller Mode */
+	DL_I2C_resetControllerTransfer(config->base);
+
+	/* Config other settings */
+	DL_I2C_setControllerTXFIFOThreshold(config->base, DL_I2C_TX_FIFO_LEVEL_BYTES_1);
+	DL_I2C_setControllerRXFIFOThreshold(config->base, DL_I2C_RX_FIFO_LEVEL_BYTES_1);
+	DL_I2C_enableControllerClockStretching(config->base);
+
+	/* Configure Interrupts */
+	DL_I2C_clearInterruptStatus(config->base, TI_MSPM0_CONTROLLER_INTERRUPTS);
+	DL_I2C_enableInterrupt(config->base, TI_MSPM0_CONTROLLER_INTERRUPTS);
+
+	/* Enable module */
+	DL_I2C_enableController(config->base);
+
+	return ret;
+}
+
+static int i2c_mspm0_receive(const struct device *dev, struct i2c_msg msg, uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+
+	/* Update cached msg and addr */
+	data->msg = msg;
+	data->addr = addr;
+
+	/* Send a read request to Target */
+	data->count = 0;
+	data->state = I2C_MSPM0_RX_STARTED;
+	DL_I2C_startControllerTransfer(config->base, data->addr, DL_I2C_CONTROLLER_DIRECTION_RX,
+				       data->msg.len);
+
+	while (!(DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_IDLE)) {
+	}
+
+	if (data->state == I2C_MSPM0_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	/* If error, return error */
+	if (((DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_ERROR) != 0x00) ||
+	    (data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transmit(const struct device *dev, struct i2c_msg msg, uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+
+	/* Update cached msg, addr, and state */
+	data->msg = msg;
+	data->addr = addr;
+	data->state = I2C_MSPM0_IDLE;
+
+	/* Flush anything that is left in the stale FIFO */
+	DL_I2C_flushControllerTXFIFO(config->base);
+
+	/*
+	 *  Fill the FIFO
+	 *  The FIFO is 8-bytes deep, and this function will return number
+	 *  of bytes written to FIFO
+	 */
+	data->count = DL_I2C_fillControllerTXFIFO(config->base, data->msg.buf, data->msg.len);
+
+	/* Enable TXFIFO trigger interrupt if there are more bytes to send */
+	if (data->count < data->msg.len) {
+		DL_I2C_enableInterrupt(config->base, DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	} else {
+		DL_I2C_disableInterrupt(config->base, DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	}
+
+	/*
+	 * Send the packet to the controller.
+	 * This function will send Start + Stop automatically
+	 */
+	data->state = I2C_MSPM0_TX_STARTED;
+	while (!(DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_IDLE)) {
+	}
+
+	DL_I2C_startControllerTransfer(config->base, data->addr, DL_I2C_CONTROLLER_DIRECTION_TX,
+				       data->msg.len);
+
+	while (!(DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_IDLE)) {
```

Comment:
Tight polling loops like this aren't inherently wrong but typically there's an interrupt signal available.

Would recommend changing the driver, if possible, to use an interrupt (event driven) state machine that is kicked off by the transfer function, and completed when the state machine signals back to the thread that had called transfer blocking on a k_sem_take by giving a semaphore.

If you really want the tight poll loop here instead, or there's no other option, use a WAIT_FOR() macro with a timeout such that even given a faulty i2c transfer the caller isn't blocked forever and instead may get an error.

https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#ga68eb35df6b4715714312df90209accee